src/main/java/de/waitpermission/joppen/CoreFormatter.java

```diff
@@ -0,0 +1,59 @@
+package de.waitpermission.joppen;
+
+import java.io.Writer;
+
+/**
+ * The core formatting algorithm that turns a sequence of tokens into string output. This implements
+ * an algorithm along the lines of that in section 4 of Oppen's paper.
+ */
+public final class CoreFormatter {
+
+  /** This is where the output gets written to. */
+  private final Writer sink;
```

Comment:
In hindsight, `Appendable` is probably a better choice here.

Reply:
Yep, I agree